internal/task: remove coroutines

niaow opened the pull request.

niaow commented:
This is a bit bigger than I expected, so I don't think it is necessarily a good idea to include this for this release. Next release cycle I would like to refactor a lot of the things that used to depend on coroutines.

aykevl commented on compileopts/config.go:
```diff
@@ -159,31 +159,6 @@ func (c *Config) OptLevels() (optLevel, sizeLevel int, inlinerThreshold uint) {
-		// being pointed to doesn't need a context. The function pointer is a
-		// regular function pointer.
-		return "doubleword"
-	case "none", "coroutines":
-		// As "doubleword", but with the function pointer replaced by a unique
-		// ID per function signature. Function values are called by using a
-		// switch statement and choosing which function to call.
-		// Pick the switch implementation with the coroutines scheduler, as it
-		// allows the use of blocking inside a function that is used as a func
-		// value.
-		return "switch"
```
FYI: this does have an effect on AVR which used `-scheduler=none` by default. This used to be necessary because function pointers used to be buggy in AVR but I believe it should work now (except a very small increase in binary size sometimes which I think is acceptable).

aykevl reviewed (commented):
Phew! That's indeed a big PR. I did expect it to be pretty big though...

I think I would have tried to split this into multiple commits (remove coroutines scheduler, remove switch implementation of functions, and then remove the `parentHandle` parameter). But I'm not going to complain.

We might want to test AVR some more just to be sure, see below.

Anyway, giving a preliminary LGTM now.

niaow commented:
> I think I would have tried to split this into multiple commits

Yeah, I can try to do that

niaow committed:
internal/task: remove coroutines

niaow committed:
transform: remove switched func lowering

The switched func lowering was mainly necessary for coroutines.
With coroutines removed, this is no longer necessary.

niaow committed:
compiler: remove parentHandle from calling convention

This removes the parentHandle argument from the internal calling convention.
It was formerly used to implment coroutines.
Now that coroutines have been removed, it is no longer necessary.

niaow commented:
This PR has now been split into 3 less-huge commits.

aykevl reviewed (approved):
Looks good to me!
Feel free to fix (probably trivial) merge conflicts and merge the PR.

niaow commented:
I need to look into why there is an issue with the closure on Windows.

niaow commented:
It seems to be a random error. . . which is concerning. I also distinctly remember running into that exact error before (so I don't think it is caused by this PR). I strongly suspect that the actual fix is #2433, which I now realize I forgot to finish debugging.

niaow commented:
I created #2551 for now

niaow commented:
IDK, does anyone object to merging this now?

dkegel-fastly commented:
I'm all for merging it if it will prevent bad user experiences.  We can always resurrect coroutines later...
